src/Swarm/Game/ScenarioInfo.hs

```diff
@@ -126,85 +134,71 @@ flatten (SISingle p) = [p]
 flatten (SICollection _ c) = concatMap flatten $ scenarioCollectionToList c
 
 -- | Load all the scenarios from the scenarios data directory.
-loadScenariosWithWarnings ::
+loadScenarios ::
+  (Has (Accum (Seq SystemFailure)) sig m, Has (Lift IO) sig m) =>
   EntityMap ->
-  IO ([SystemFailure], ScenarioCollection)
-loadScenariosWithWarnings em = do
-  res <- getDataDirSafe Scenarios "scenarios"
+  m ScenarioCollection
+loadScenarios em = do
+  res <- runThrow @SystemFailure $ getDataDirSafe Scenarios "scenarios"
   case res of
-    Left err -> return ([err], SC mempty mempty)
+    Left err -> do
+      warn err
+      return $ SC mempty mempty
     Right dataDir -> loadScenarioDir em dataDir
 
 -- | The name of the special file which indicates the order of
 --   scenarios in a folder.
 orderFileName :: FilePath
 orderFileName = "00-ORDER.txt"
 
-readOrderFile :: (MonadIO m) => FilePath -> m [String]
+readOrderFile :: (Has (Lift IO) sig m) => FilePath -> m [String]
 readOrderFile orderFile =
-  filter (not . null) . lines <$> liftIO (readFile orderFile)
+  filter (not . null) . lines <$> sendIO (readFile orderFile)
 
 -- | Recursively load all scenarios from a particular directory, and also load
 --   the 00-ORDER file (if any) giving the order for the scenarios.
 loadScenarioDir ::
-  (MonadIO m) =>
+  (Has (Accum (Seq SystemFailure)) sig m, Has (Lift IO) sig m) =>
   EntityMap ->
   FilePath ->
-  m ([SystemFailure], ScenarioCollection)
+  m ScenarioCollection
 loadScenarioDir em dir = do
   let orderFile = dir </> orderFileName
       dirName = takeBaseName dir
-  orderExists <- liftIO $ doesFileExist orderFile
+  orderExists <- sendIO $ doesFileExist orderFile
   morder <- case orderExists of
     False -> do
-      when (dirName /= "Testing") $
-        liftIO . putStrLn $
-          "Warning: no "
-            <> orderFileName
-            <> " file found in "
-            <> dirName
-            <> ", using alphabetical order"
+      when (dirName /= "Testing") . warn $
+        OrderFileWarning (dirName </> orderFileName) NoOrderFile
       return Nothing
     True -> Just <$> readOrderFile orderFile
-  itemPaths <- liftIO $ keepYamlOrPublicDirectory dir =<< listDirectory dir
+  itemPaths <- sendIO $ keepYamlOrPublicDirectory dir =<< listDirectory dir
 
   case morder of
     Just order -> do
       let missing = itemPaths \\ order
           dangling = order \\ itemPaths
 
-      unless (null missing) $
-        liftIO . putStr . unlines $
-          ( "Warning: while processing "
-              <> (dirName </> orderFileName)
-              <> ": files not listed in "
-              <> orderFileName
-              <> " will be ignored"
-          )
-            : map ("  - " <>) missing
+      unless (null missing) . warn $
+        OrderFileWarning (dirName </> orderFileName) (MissingFiles missing)
```

Comment:
Here, we can exploit the stronger `MissingFiles (NonEmpty FilePath)` type:
```suggestion
      forM_ (NE.nonEmpty missing) $ warn .
        OrderFileWarning (dirName </> orderFileName) . MissingFiles
```

Reply:
Nice idea!